Modules/Filtering/FFT/test/itkComplexToComplex1DFFTImageFilterTest.cxx

```diff
@@ -110,12 +112,21 @@ itkComplexToComplex1DFFTImageFilterTest(int argc, char * argv[])
 #endif
     using FFTInverseType = itk::ComplexToComplex1DFFTImageFilter<ComplexImageType, ComplexImageType>;
     auto inverse = FFTInverseType::New();
+
     if (inverse == nullptr)
     {
       std::cerr << "Failed to register a backend for ComplexToComplex1DFFTImageFilter" << std::endl;
       return EXIT_FAILURE;
     }
 
+#ifndef ITK_FFT_FACTORY_REGISTER_MANAGER
+#  if defined(ITK_USE_FFTWD) || defined(ITK_USE_FFTWF)
+    ITK_EXERCISE_BASIC_OBJECT_METHODS(inverse, FFTWComplexToComplex1DFFTImageFilter, ImageToImageFilter);
+#  else
+    ITK_EXERCISE_BASIC_OBJECT_METHODS(inverse, VnlComplexToComplex1DFFTImageFilter, ImageToImageFilter);
+#  endif
+#endif
```

Comment:
Following from the previous comment, `inverse` here is an `itk::ComplexToComplex1DFFTImageFilter` with some backend returned from the object factory. We could try downcasting to each backend and exercising object methods, but it's easier to just use the base FFT class:

```cpp
ITK_EXERCISE_BASIC_OBJECT_METHODS(inverse, ComplexToComplex1DFFTImageFilter, ImageToImageFilter);
```

Also note that `#ifndef ITK_FFT_FACTORY_REGISTER_MANAGER` isn't necessary here as `ITK_FFT_FACTORY_REGISTER_MANAGER` will never be defined for ITK tests. #3002 will streamline default factory registration in the test kernel.

Reply:
After building and experimenting a bit more it looks like the solution may be a bit more complicated than I initially expected. For an object factory override of `ComplexToComplexFFTImageFilter` each of the following calls returns a different name:

```
inverseComplexFilter->GetNameOfClass() -->VnlComplexToComplexFFTImageFilter
inverseComplexFilter->Self::GetNameOfClass() -->ComplexToComplexFFTImageFilter
inverseComplexFilter->Superclass::GetNameOfClass() -->ImageToImageFilter
```

This behavior is expected but a bit opaque as `GetNameOfClass` is overridden via object factory generation. I wonder if it would be pragmatic to add an additional test macro that calls `->Self::GetNameOfClass()` rather than `->Superclass::GetNameOfClass()` so that we can check the base FFT name and ignore the specific implementation?

Reply:
I was not being able to get a clear picture out of the previous explanations, but your last comment 
https://github.com/InsightSoftwareConsortium/ITK/pull/3003#discussion_r776032896 seems to match what I was seeing when testing locally.

I think the current commit b36f5a8 will pass, as I tested it locally.

I am not sure all this is due to factories or backends, so not sure how a new eventual macro should be named to test these cases. And I'm not sure whether we can ensure that in all cases the hierarchy will have at most 3 levels. Maybe downcasting could be used maybe to call those `itkTypeMacro`s that we chase here. Otherwise, if you think that a new macro is necessary/handy for these cases/can better test these in the future, +1 for it.

As for the comment about `#ifndef ITK_FFT_FACTORY_REGISTER_MANAGER` I really did not think much about it, only re-used what was in place when registering the factories.

In either case, IMHO whether testing through downcasting, adding a new macro, and removing the `#ifndef ITK_FFT_FACTORY_REGISTER_MANAGER` guard would be best addressed by subsequent, separate PRs if the current commit's CIs come green.

Reply:
~I think the current commit b36f5a8 will pass, as I tested it locally.~

Failed again :man_facepalming:, so I am commenting the `ITK_EXERCISE_BASIC_OBJECT_METHODS` call. I think best is to address it separately.

Reply:
Documented thoughts and potential resolutions in #3008 , feel free to continue discussion there

Reply:
:+1: Thanks @tbirdso.

Reply:
FYI #3010 addresses the issue with name checking upcasted classes with `ITK_EXERCISE_BASIC_OBJECT_METHODS`. Fine to either uncomment here or push to a subsequent PR.

Reply:
Rebasing on `master` and uncommenting the tests calls if preferred.

Reply:
Done in 5643fdb. The concerned tests are now passing locally. Thanks for the time given to this @tbirdso.